core/src/main/java/org/elasticsearch/indices/cluster/IndicesClusterStateService.java

```diff
@@ -432,114 +445,98 @@ private void updateIndices(ClusterChangedEvent event) {
                     createShard(nodes, routingTable, shardRouting, indexService);
                 } else {
-                    updateShard(event, nodes, shardRouting, shard);
+                    updateShard(nodes, shardRouting, shard);
                 }
             }
         }
     }
 
     private void createShard(DiscoveryNodes nodes, RoutingTable routingTable, ShardRouting shardRouting,
                              AllocatedIndex<? extends Shard> indexService) {
-        assert shardRouting.initializing();
+        assert shardRouting.initializing() : "only allow shard creation for initializing shard but was " + shardRouting;
 
-        if (shardRouting.initializing()) {
-            DiscoveryNode sourceNode = null;
-            if (shardRouting.isPeerRecovery()) {
-                sourceNode = findSourceNodeForPeerRecovery(logger, routingTable, nodes, shardRouting);
-                if (sourceNode == null) {
-                    logger.trace("ignoring initializing shard {} - no source node can be found.", shardRouting.shardId());
-                    return;
-                }
+        DiscoveryNode sourceNode = null;
+        if (shardRouting.isPeerRecovery()) {
+            sourceNode = findSourceNodeForPeerRecovery(logger, routingTable, nodes, shardRouting);
+            if (sourceNode == null) {
+                logger.trace("ignoring initializing shard {} - no source node can be found.", shardRouting.shardId());
+                return;
             }
+        }
 
-            try {
-                logger.debug("{} creating shard", shardRouting.shardId());
-                RecoveryState recoveryState = recoveryState(nodes.getLocalNode(), sourceNode, shardRouting,
-                    indexService.getIndexSettings().getIndexMetaData());
-                indicesService.createShard(shardRouting, recoveryState, recoveryTargetService, new RecoveryListener(shardRouting),
-                    repositoriesService, nodeServicesProvider, failedShardHandler);
-            } catch (IndexShardAlreadyExistsException e) {
-                // ignore this, the method call can happen several times
-                assert false;
-            } catch (Throwable e) {
-                failAndRemoveShard(shardRouting, true, "failed to create shard", e);
-            }
+        try {
+            logger.debug("{} creating shard", shardRouting.shardId());
+            RecoveryState recoveryState = recoveryState(nodes.getLocalNode(), sourceNode, shardRouting,
+                indexService.getIndexSettings().getIndexMetaData());
+            indicesService.createShard(shardRouting, recoveryState, recoveryTargetService, new RecoveryListener(shardRouting),
+                repositoriesService, nodeServicesProvider, failedShardHandler);
+        } catch (IndexShardAlreadyExistsException e) {
+            // ignore this, the method call can happen several times
+            assert false;
```

Comment:
This is code I want to remove as I think this situation is impossible to happen (only ICSS does create a shard and it checks before that the shard does not exist). As part of this PR, I've just added the assert to give it a few cycles on CI before I remove this exception altogether.